### PR TITLE
Add support for SGP41

### PR DESCRIPTION
**Pull request by martgras**
# What does this implement/fix?

add support for SPG41 and SPG40. SGP41 supports all features of SGP40 but measures NOx as well.
The sensor type is autodetected.
Algorithm tuning option added.

Because spg41 adds NOx the new component is a multi-sensor  component therefore a  new configuration schema is required.
Therefore I decided to create a new component SPG4x supporting both sensors and mark the existing SPG40 component deprecated.  Just changing the SGP40 component would break existing setups 

Instead of copying the sensirion gas library code I'm now adding it as library. (https://github.com/Sensirion/arduino-gas-index-algorithm)
~~The only "problem" is esp-idf support.  Because sensirion doesn't supply library.json the lib is considered arduino only.  But actually it is framework agnostic.  I submitted https://github.com/Sensirion/arduino-gas-index-algorithm/pull/1 but in the meantime use my own fork at  https://github.com/martgras/arduino-gas-index-algorithm .
Works like a charm with esp-idf~~

All other code is using the codebase from SGP40 by @SenexCrenshaw

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2016

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
  - platform: sgp4x
    voc:
      name: "VOC Index"
      id: sgp41_voc_index
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    nox:
      name: "NOx"
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
        
    compensation:
      humidity_source: sht40_humidity
      temperature_source: sht40_temperature
    update_interval: ${updates}


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs/).

### CODE REVIEW

File: esphome/components/sgp4x/sensor.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                    cfg[CONF_GATING_MAX_DURATION_MINUTES],
+                    cfg[CONF_GAIN_FACTOR],
+                )
+            )
+
+    # if https://github.com/Sensirion/arduino-gas-index-algorithm/pull/1 gets mergged it can be replaced with
+    #    cg.add_library("sensirion/Sensirion Gas Index Algorithm", "^3.1.0")
+    # Currently the library is missing a library.json manifest and is therefore considered arduino only by pio but actually is library agnostic
+    # The fork only adds the missing manifest for esp-idf compability
+    cg.add_library("https://github.com/martgras/arduino-gas-index-algorithm.git", "")
```
Comment (martgras) [inline esphome/components/sgp4x/sensor.py]: Should this repo move to esphome?
Don't know if sensirion reacts to my PR

**Comment by jesserockz** [comment]
They will need to make another release on platformio for this to work. But for now you could point `platformio.ini` and `cg.add_library` to their repo.

File: esphome/components/sgp4x/sensor.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            cg.add(
+                var.set_nox_algorithm_tuning(
+                    cfg[CONF_INDEX_OFFSET],
+                    cfg[CONF_LEARNING_TIME_OFFSET_HOURS],
+                    cfg[CONF_LEARNING_TIME_GAIN_HOURS],
+                    cfg[CONF_GATING_MAX_DURATION_MINUTES],
+                    cfg[CONF_GAIN_FACTOR],
+                )
+            )
+    cg.add_library("https://github.com/Sensirion/arduino-gas-index-algorithm.git", "")
```
Comment (jesserockz) [inline esphome/components/sgp4x/sensor.py]:
```suggestion
    cg.add_library(None, None, "https://github.com/Sensirion/arduino-gas-index-algorithm.git", "")
```

File: platformio.ini [hunk 3]
```diff
@@ -39,6 +39,7 @@ lib_deps =
     bblanchon/ArduinoJson@6.18.5           ; json
     wjtje/qr-code-generator-library@1.7.0  ; qr_code
     functionpointer/arduino-MLX90393@1.0.0 ; mlx90393
+    https://github.com/Sensirion/arduino-gas-index-algorithm.git ; Sensirion Gas Index Algorithm Arduino Library
```
Comment (jesserockz) [inline platformio.ini:43]:
```suggestion
    ; This is using the repository until a new release is published to PlatformIO
    https://github.com/Sensirion/arduino-gas-index-algorithm.git ; Sensirion Gas Index Algorithm Arduino Library
```

**Review by jesserockz — APPROVED** [review]
Minor changes, otherwise looks fine.

**Pull request merged**